pkg/daemon/daemon.go

```diff
@@ -750,18 +764,84 @@ func (dn *Daemon) annotateNode(node, value string) error {
 func (dn *Daemon) drainNode(name string) error {
 	glog.Info("drainNode(): Update prepared")
 	var err error
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
 	// wait a random time to avoid all the nodes drain at the same time
-	wait.PollUntil(time.Duration(rand.Intn(15)+1)*time.Second, func() (bool, error) {
+	time.Sleep(wait.Jitter(3*time.Second, 3))
+	wait.JitterUntil(func() {
 		if !dn.drainable {
-			glog.Info("drainNode(): other node is draining, waiting...")
+			glog.V(2).Info("drainNode(): other node is draining")
+			return
 		}
-		return dn.drainable, nil
-	}, dn.stopCh)
+		glog.V(2).Info("drainNode(): no other node is draining")
+		err = dn.annotateNode(dn.name, annoDraining)
+		if err != nil {
+			glog.Errorf("drainNode(): Failed to annotate node: %v", err)
+			return
+		}
+		cancel()
+	}, 3*time.Second, 3, true, ctx.Done())
 
-	err = dn.annotateNode(dn.name, annoDraining)
-	if err != nil {
-		glog.Errorf("drainNode(): Failed to annotate node: %v", err)
-		return err
+	if utils.ClusterType == utils.ClusterTypeOpenshift {
+		mcpInformerFactory := mcfginformers.NewSharedInformerFactory(dn.mcClient,
+			time.Second*30,
+		)
+		mcpInformer := mcpInformerFactory.Machineconfiguration().V1().MachineConfigPools().Informer()
+
+		mcpList, err := dn.mcClient.MachineconfigurationV1().MachineConfigPools().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			glog.Errorf("drainNode(): failed to list Machine Config Pools: %v", err)
+			return err
+		}
+		var mcp mcfgv1.MachineConfigPool
+		for _, mcp = range mcpList.Items {
+			selector, err := metav1.LabelSelectorAsSelector(mcp.Spec.NodeSelector)
+			if err != nil {
+				glog.Errorf("drainNode(): Machine Config Pool %s invalid label selector: %v", mcp.GetName(), err)
+				return err
+			}
+
+			if selector.Matches(labels.Set(dn.node.Labels)) {
+				dn.mcpName = mcp.GetName()
+				glog.Infof("drainNode(): find node in MCP %s", dn.mcpName)
+				break
+			}
+		}
+
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
+
+		mcpEventHanedler := func(obj interface{}) {
+			newMcp := obj.(*mcfgv1.MachineConfigPool)
+			if newMcp.GetName() != dn.mcpName {
+				return
+			}
+			if mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolDegraded) &&
+				mcfgv1.IsMachineConfigPoolConditionTrue(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdated) &&
+				mcfgv1.IsMachineConfigPoolConditionFalse(newMcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdating) {
+				glog.Infof("drainNode(): MCP %s is ready", newMcp.GetName())
```

Comment:
Check `readyMachineCount` == `machineCount` as well ?

Reply:
I suppose the 'Updated' condition of MCP is sufficient.